# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [6]:
# 1. Гіпотези
times = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7   # середній час до покупки історично
alpha = 0.1

# Нульова гіпотеза H0: mu0 = 123.7 
# Адьтернативна гіротеза H1: mu > 123.7 

# 2. Обчислимо вибіркову статистику
x_bar = np.mean(times)
s = np.std(times, ddof=1)   # стандартне відхилення з n-1
n = len(times)

In [15]:
# 3. Правосторонній тест
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

df = n - 1
t_crit = stats.t.ppf(alpha, df)

p_value = 1- stats.t.cdf(t_stat, df)

In [16]:
print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"p-value: {p_value:.4f}")


Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
t-статистика: 2.069
Критичне значення (df=6): -1.440
p-value: 0.0420


In [11]:
# scipy тест
res = stats.ttest_1samp(times, popmean=mu0, alternative="greater")
print(f"scipy: t = {res.statistic:.3f}, p-value = {res.pvalue:.5f}")

scipy: t = 2.069, p-value = 0.04202


In [17]:
# 4. висновок на основі р-value
if res.pvalue < alpha:
    print(f"Відхиляємо H0: середній час до покупки більший за 123.7 на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: середній час до покупки більший за 123.7 на рівні значущості 0.1


5. За наявними даними є підстави вважати, що редизайн сайту подовжив час до покупки.
Ми отримали:
x=132.29, mu0=123.7,t≈2.07,p-value≈0.042

Оскільки для перевірки гіпотези:
H0 :mu0=123.7,H1:mu>123.7
маємо
p-value=0.042<α
то нульову гіпотезу відхиляємо.
До того ж t-статситика є теж більшою за критичне значення 
t_stat > t_crit
Висновок
На основі вибірки є статистично значущі підстави вважати, що після редизайну середній час до покупки збільшився.

6. Заміна alpha = 0.5
   При α=0.05 висновок теж той самий, бо p-value залишається незмінним
 0.042 < 0.05

Тобто і при 5%, і при 10% рівні значущості висновок однаковий: редизайн статистично значущо подовжив час до покупки.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [18]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

In [20]:
count = 311   # кількість "успіхів" - логотип компаніі впізнали
nobs = 1200    # обсяг вибірки
p0 = 0.23      # гіпотетична пропорція (H0)

# z-тест для пропорції
stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha=0.03, method='normal')

print("🔹 Тест для однієї пропорції")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")

🔹 Тест для однієї пропорції
Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


Проведено односторонній тест
1. Односторонній тест

   * $H_0: p = 0.23$
   * $H_1: p > 0.23$
   * Оскільки $p\text{-value} = 0.0106 < 0.03$, відхиляємо $H_0$.
     👉 Є статистично значущі докази, що частка тих, хто впізнав доготип після рекламної кампанії, **більша за 23%**.

2. 97% довірчий інтервал (0.2317, 0.2866)

   * Це означає, що з імовірністю 97% справжня частка лежить у межах **від 23.17% до 28.66%**.
Цей інтервал означає, що правдоподібні значення істинної частки впізнавання після кампанії лежать приблизно між 23.17% і 28.66%.
Оскільки весь інтервал лежить вище 23%, це узгоджується з нашим тестом: дані підтримують висновок, що справжня частка стала більшою за 23%.